src/Raven.Server/Documents/ETL/Providers/AI/GenAi/GenAiScriptTransformer.cs

```diff
@@ -91,35 +154,95 @@ public override void Transform(GenAiItem item, GenAiStatsScope stats, EtlProcess
-        _stats.NumberOfContextObjects++;
 
-        string hash = CalculateHash(context);
-        var isCached = ShouldSendContext(hash, _configuration.Identifier, Current.Document) == false;
+    private void ProcessScriptResults()
+    {
+        ObjectInstance ai = DocumentScript.ScriptEngine.GetValue("ai").AsObject();
+        Function retrieveContexts = ai.Prototype!.GetOwnProperty("__retrieveContexts").Value.AsFunctionInstance();
+        JsArray contexts = retrieveContexts.Call(ai, []).AsArray();
+        List<string> attachmentsHashes = null;
+        foreach (var ctx in contexts)
+        {
+            attachmentsHashes?.Clear();
+            
+            ObjectInstance ctxObj = ctx.AsObject();
+            ObjectInstance userSpecificCtx = ctxObj.GetOwnProperty("ctx").Value.AsObject();
+            var context = JsBlittableBridge.Translate(Context, DocumentScript.ScriptEngine, userSpecificCtx);
+            _stats.NumberOfContextObjects++;
+
+            foreach (var current in ctxObj.GetOwnProperty("attachments").Value.AsArray())
+            {
+                attachmentsHashes ??= [];
+                
+                var attachmentObj = current.AsObject();
+                var data = attachmentObj.GetOwnProperty("data").Value;
+                if (data.IsNull() == false)
+                    attachmentsHashes.Add(_attachments?.TryGetValue(data, out var attachment) is true ? attachment?.Base64Hash.ToString() : data.AsString());
+
+                attachmentsHashes.Add(attachmentObj.GetOwnProperty("type").Value.AsString());
+            }
+            
+            string hash = CalculateHash(context, attachmentsHashes);
+            var isCached = ShouldSendContext(hash, _configuration.Identifier, Current.Document) == false;
 
-        if (isCached)
-            _stats.TotalCachedContexts++;
+            if (isCached)
+                _stats.TotalCachedContexts++;
 
-        using (context)
-        {
-            _currentRun.Add(new GenAiScriptResult(Current.DocumentId, context.CloneOnTheSameContext(), hash, isCached));
+            using (context)
+            {
+                var result = new GenAiScriptResult(Current.DocumentId, context.CloneOnTheSameContext(), hash, isCached);
+                if (attachmentsHashes?.Count > 0)
+                {
+                    result.Attachments = [];
+                    
+                    foreach (var current in ctxObj.GetOwnProperty("attachments").Value.AsArray())
+                    {
+                        var attachmentObj = current.AsObject();
+                        var reference = attachmentObj.GetOwnProperty("data").Value;
+                        var data = string.Empty;
+                        string type = attachmentObj.GetOwnProperty("type").Value.AsString();
+                        string filename = "unknown.name";
+
+                        // TODO: we aren't being really efficient here in terms of allocations / memory
```

Comment:
my concern is that practically speaking this approach has some limitations and is not supported for AI Assistant. At the end we will have to develop Files API support and will end up with 2 different implementations anyway.

Reply:
I retract my concern. Files API seems to be limited at the moment as well. It looks like only PDFs are supported for the chat/completions.
I did a really dirty POC available here: https://github.com/ppekrol/ravendb/tree/feature/files-api

Reply:
Also Ollama does not support that at all :)